### PR TITLE
Enable razor UTs in-memory compilation

### LINKED ISSUE #9010: Enable razor UTs in-memory compilation
- [ ] Link all the steps to allow razor UTs to run with in-memory compilation.
- [ ] Fix [razor symbol reference](https://trello.com/c/a1wUeO5j/1319-fix-razor-uts-for-symbol-references-2devs-1w) issue
- [ ] Final cleanup:
  - There's probably a build locator (or other) dependency that will go away
  - Verifier `IsRazor` property and `CompileRazor`, `LangVersion`, `PrepareRazorProject`, `PrepareRazorFiles` and `RestorePackages` methods will go away

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #9010

This includes the previously reverted PR #9190 and additional fixes that were needed to make it work.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/FileMetadataAnalyzerTest.cs [hunk 1]
```diff
@@ -106,13 +106,12 @@ public void CreateMessage_NoEncoding_SetsEmptyString(bool isTestProject)
...
-        [DataRow("Razor.cshtml", "EmptyProject.GlobalUsings.g.cs", ".NETCoreApp,Version=v7.0.AssemblyAttributes.cs", "EmptyProject.AssemblyInfo.cs", "EmptyProject.RazorAssemblyInfo.cs")]
-        public void Verify_RazorFilesAreIgnored(string fileName, params string[] expectedFiles) =>
+        [DataRow("Razor.razor")]
+        [DataRow("Razor.cshtml")]
+        public void Verify_RazorFilesAreIgnored(string fileName) =>
             CreateBuilder(ProjectType.Product, fileName)
-                .WithAdditionalFilePath(AnalysisScaffolding.CreateSonarProjectConfig(TestContext, ProjectType.Product))
                 .VerifyUtilityAnalyzer<FileMetadataInfo>(messages =>
-                    messages.Select(x => Path.GetFileName(x.FilePath)).Should().Contain(expectedFiles));    // There are more files on some PCs: JSExports.g.cs, LibraryImports.g.cs, JSImports.g.cs
+                    messages.Select(x => Path.GetFileName(x.FilePath)).Should().BeEmpty());    // There are more files on some PCs: JSExports.g.cs, LibraryImports.g.cs, JSImports.g.cs
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/FileMetadataAnalyzerTest.cs:114]:
```suggestion
                    x.Select(fileInfo => Path.GetFileName(fileInfo.FilePath)).Should().BeEmpty());    // There are more files on some PCs: JSExports.g.cs, LibraryImports.g.cs, JSImports.g.cs
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/CollectionPropertiesShouldBeReadOnly.razor.cs [hunk 2]
```diff
@@ -1,11 +1,13 @@
 ﻿using Microsoft.AspNetCore.Components;
+using System.Collections.Generic;
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/CollectionPropertiesShouldBeReadOnly.razor.cs:2]: Unused using

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/SymbolReferenceAnalyzerTest.cs [hunk 3]
```diff
@@ -223,32 +223,32 @@ public void Verify_UnchangedFiles(string unchangedFileName, bool expectedProtobu
...
+                        VerifyReferences(orderedSymbols[0].Reference, 10, 43);                      // y
+                        VerifyReferences(orderedSymbols[0].Reference, 10, 45, 42);                  // LocalMethod
+                        VerifyReferences(orderedSymbols[0].Reference, 10, 58, 52);                  // AdditionalAttributes
+
+                        VerifyReferencesColumns(orderedSymbols[0].Reference, 14, 5, 5, 19, 31);     // currentCount: line 5
+                        VerifyReferencesColumns(orderedSymbols[0].Reference, 14, 7, 7, 1, 13);      // currentCount: line 7
+                        VerifyReferencesColumns(orderedSymbols[0].Reference, 14, 21, 21, 8, 20);    // currentCount: line 21
+                        VerifyReferencesColumns(orderedSymbols[0].Reference, 14, 49, 49, 26, 38);   // currentCount: line 49
+
+                        orderedSymbols[1].FilePath.Should().EndWith("RazorComponent.razor");        // RazorComponent.razor
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/SymbolReferenceAnalyzerTest.cs:245]: I wonder why the order has changed between `RazorComponent.razor` and `Razor.razor.cs`

File: analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs [hunk 4]
```diff
@@ -121,15 +124,12 @@ private string GetTestCaseFileRelativePath(string path)
...
     private bool IsExtensionOfSupportedType(FileInfo fileInfo) =>
         fileInfo.Extension.Equals(fileExtension, StringComparison.OrdinalIgnoreCase)
-        || fileInfo.Extension.Equals(".razor", StringComparison.OrdinalIgnoreCase)
+        || (fileInfo.Extension.Equals(".razor", StringComparison.OrdinalIgnoreCase) && project.Language == LanguageNames.CSharp)
         || (fileInfo.Extension.Equals(".cshtml", StringComparison.OrdinalIgnoreCase) && project.Language == LanguageNames.CSharp);
 
     private static ProjectBuilder AddDocument(Project project, string fileName, string fileContent) =>
         FromProject(project.AddDocument(fileName, fileContent).Project);
 
-    private static ProjectBuilder AddAdditionalDocument(Project project, string fileName, string fileContent) =>
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.TestFramework/Build/ProjectBuilder.cs:130]: Why was this moved?

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs [hunk 5]
```diff
@@ -19,15 +19,12 @@
  */
 
 using System.IO;
+using System.Linq;
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs:22]: Not used

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs [hunk 6]
```diff
@@ -84,17 +82,13 @@ public Verifier(VerifierBuilder builder)
...
         }
-        if (builder.IsRazor)
-        {
-            if (builder.ParseOptions.IsEmpty)
-            {
-                throw new ArgumentException($"{nameof(builder.IsRazor)} was set. {nameof(ParseOptions)} must be specified.");
-            }
-            else if (language != AnalyzerLanguage.CSharp)
+        razorFilePaths = builder.Paths.Select(TestCasePath).Where(IsRazorOrCshtml)
+            .Concat(builder.Snippets.Where(x => IsRazorOrCshtml(x.FileName)).Select(x =>
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs:86]:
```suggestion
        razorFilePaths = builder.Paths
            .Select(TestCasePath)
            .Where(IsRazorOrCshtml)
            .Concat(builder.Snippets.Where(x => IsRazorOrCshtml(x.FileName)).Select(x =>
            ...
            .ToArray();
```

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs [hunk 7]
```diff
@@ -84,17 +82,13 @@ public Verifier(VerifierBuilder builder)
...
             {
-                throw new ArgumentException($"{nameof(builder.IsRazor)} was set for {language} analyzer. Only C# is supported.");
-            }
-        }
+                var tempFilePath = Path.Combine(Directory.GetCurrentDirectory(), "TestCases", x.FileName);
+                File.WriteAllText(tempFilePath, x.Content);
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/Verifier.cs:89]: Please leave a comment here explaining why they cannot be in-memory

**Review by CristianAmbrosini — CHANGES_REQUESTED** [review]
Left a few ideas/questions

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/CopyPasteTokenAnalyzerTest.cs [hunk 8]
```diff
@@ -20,176 +20,171 @@
...
+                });
 
 #if NET
 
-        [DataTestMethod]
-        [DataRow("Razor.razor")]
-        [DataRow("Razor.cshtml")]
-        public void Verify_NoMetricsAreComputedForRazorFiles(string fileName) =>
-            CreateBuilder(ProjectType.Product, fileName)
-                .WithAdditionalFilePath(AnalysisScaffolding.CreateSonarProjectConfig(TestContext, ProjectType.Product))
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/CopyPasteTokenAnalyzerTest.cs:141]: No. The project config is only needed for reporting issues via `SonarReportingContextBase.ReportIssueCore`.

File: analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/SymbolReferenceAnalyzerTest.cs (hunk 3 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.Test/Rules/Utilities/SymbolReferenceAnalyzerTest.cs:245]: Indeed surprising.

File: analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs [hunk 9]
```diff
@@ -73,7 +73,7 @@ public class UnnecessaryUsingsTest
         [DataRow("RandomFile_Imports.cshtml")]
         [DataRow("_Imports.cshtml")]
         public void UnnecessaryUsings_RazorViewImportsSimilarCshtmlFile_IssuesReported(string fileName) =>
-            builder
+            builder.AddReferences(NuGetMetadataReference.SystemTextJson("7.0.4"))
```
Comment (CristianAmbrosini) [inline analyzers/tests/SonarAnalyzer.Test/Rules/UnnecessaryUsingsTest.cs:76]: I would move the `AddReferences` after `AddSnippet` (same as before, also below)
